### PR TITLE
Changes to validate_docstring script to be able to check all docstrings at once

### PR DESCRIPTION
**Pull request by datapythonista**
- [ ] closes #xxxx
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This PR includes two main changes:
1. When calling the `scripts/validate_docstrings.py` script without parameters, it returns a JSON with the validation of all docstrings.
2. Refactoring the structure of the code, mainly to be able to implement 1.

The previous behavior when calling the script with parameters was to generate a csv with basic information about each docstring (not the full validation, but basic things like whether sections existed, or the file where the function is defined).

The main change is that `validate_one()` now returns the results as Python objects, instead of printing directly to the console. Few other changes to make the code cleaner are implemented (mainly moving `_load_obj` and `_original_callable` inside the `Docstring` class.

### CODE REVIEW

File: scripts/validate_docstrings.py
Comment (WillAyd) [inline scripts/validate_docstrings.py]: Is the reason why we write different output for one vs all simply console readability? Wondering if we shouldn't align the output regardless of how this is invoked (maybe a separate PR depending on complexity)

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/scripts/test_validate_docstrings.py
Comment (jorisvandenbossche) [reply pandas/tests/scripts/test_validate_docstrings.py]: In general that is true, however in this case I think it is good to test it is actually a zero-length list and not something else Falsey?

**Review by jorisvandenbossche — COMMENTED** [review]
Can you give a quick summary of the changes? (it seems more a bit of refactor than new functionality? Or are there user-visible changes?)

File: pandas/tests/scripts/test_validate_docstrings.py
Comment (datapythonista) [reply pandas/tests/scripts/test_validate_docstrings.py]: If I'm not wrong, the messages we're checking are no the whole error message, but part of it. So, this is equivalent to something like:

`assert 'error in summary' in ''.join(['there is a error in summary for this doc', 'some other error'])`

I think it makes more sense to use the `join`, as the error message will be clearer if this fails. Compared to checking all in a comprehension and them doing an `any` (which I guess will just say `False` is not `True`).

File: scripts/validate_docstrings.py
Comment (datapythonista) [reply scripts/validate_docstrings.py]: I see your point, but I think the lack of consistency is necessary in this case.

If the script is called for one object (a user validating the docstring they are changing), the "pretty" output in the console is what makes sense to return.

If the script is called for all objects (we as maintainers analyzing the status), a JSON with the information to be analyzed with pandas makes more sense than any console formatting.

Does it make sense?

File: scripts/validate_docstrings.py
Comment (datapythonista) [reply scripts/validate_docstrings.py]: I somehow agree with what you say, but I think in this case there is no way (at least I couldn't find it) to merge both use cases without increasing the amount of code and its complexity.

**Comment by datapythonista** [comment]
Thanks for the great feedback. I addressed most of the points, and commented on the others.

@jorisvandenbossche agree some more info about the PR had to be provided. I edited the description to include what's the goal, and what else has been changed.

File: scripts/validate_docstrings.py
Comment (WillAyd) [inline scripts/validate_docstrings.py]: Thanks for the docstring it helps a lot! Is there a way to make a minimal test for this? Don't want to go overboard with tests here but I am worried that if this were to somehow break in the future it could allow subtle failures in subsequent tests to go through unnoticed

**Comment by jorisvandenbossche** [comment]
> I edited the description to include what's the goal, and what else has been changed.

OK, that sounds good to me!

File: scripts/validate_docstrings.py
Comment (datapythonista) [reply scripts/validate_docstrings.py]: Makes sense to me, but I'd address it in a separate PR

**Comment by datapythonista** [comment]
@WillAyd let me know if there is any blocker to merge this

**Comment by datapythonista** [comment]
Not critical, but it'll take me few days to have the time to write those tests. And as the PR modifies the whole script, it is blocking (or will generate conflicts) with any other change to it.

#22423 is already a case, and I'd like to start creating issues for tasks in #20298, as they will save us a lot of time when reviewing the docstring updates.

But it can surely wait a couple of weeks if you think it's worth adding the tests here.

**Comment by WillAyd** [comment]
Thanks @datapythonista - I'd prefer tests up front. Hate to be a blocker on this, but realistically the longer we wait for test coverage the harder it becomes to ever get there

**Comment by datapythonista** [comment]
@WillAyd I added the requested tests, let me know if this can be merge now

**Comment by datapythonista** [comment]
@jorisvandenbossche can you take a look at this? it's the PR that makes the docstring validation script generate the list with the errors in all docstrings.

Once this is merged I'll start creating individual tickets for all the points, so we start validating all possible errors in the docstrings (and with some luck we can add it to the CI soon).

**Comment by jreback** [comment]
cc @jorisvandenbossche @WillAyd

File: scripts/tests/test_validate_docstrings.py
Comment (WillAyd) [inline scripts/tests/test_validate_docstrings.py]: Minor nit but what does this add? Can't we just do the normal inclusion check against the list instead of joining into a string?

File: scripts/tests/test_validate_docstrings.py
Comment (WillAyd) [inline scripts/tests/test_validate_docstrings.py]: Also minor but since this test isn't used for any testing (i.e. not collected by the runner) can we remove the `Test` prefix?

File: scripts/tests/test_validate_docstrings.py
Comment (WillAyd) [inline scripts/tests/test_validate_docstrings.py]: Can you use `result` instead of `res`?

File: scripts/validate_docstrings.py
Comment (WillAyd) [inline scripts/validate_docstrings.py]: Should we make the first sentence one line?

File: scripts/validate_docstrings.py
Comment (WillAyd) [inline scripts/validate_docstrings.py]: Assuming this to be the main function of the change I was thinking the test cases would cover this but don't see that - oversight on my end or something we can test?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Sorry missed the last ping

File: scripts/tests/test_validate_docstrings.py
Comment (datapythonista) [reply scripts/tests/test_validate_docstrings.py]: `msg in result['errors']` is different than this, for example `'a' in 'foo bar'` is `True`, but `a in ['foo', 'bar']` is `False`.

I think the `.join()` is simpler than another loop, is it what you would do, or you were thinking on the previous case?

Comment (WillAyd) [reply scripts/tests/test_validate_docstrings.py]: No but I thought `msg` would explicitly match one of the `errors`; will take a look more deeply on next review if that assumption is incorrect

File: scripts/validate_docstrings.py
Comment (datapythonista) [reply scripts/validate_docstrings.py]: This function is just boilerplate code, I don't think it adds any value to test that, as we would be mocking everything.

**Comment by datapythonista** [comment]
@WillAyd addressed or commented your review, can you take another look please?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks - just some more comments / detail

File: scripts/validate_docstrings.py
Comment (datapythonista) [reply scripts/validate_docstrings.py]: what exactly? the `inspect.getsourcefile()` returns the source file where a Python object is implemented. The `except TypeError` is explain in the comment below.

File: scripts/validate_docstrings.py
Comment (WillAyd) [reply scripts/validate_docstrings.py]: My mistake - thanks for clarifying

File: scripts/validate_docstrings.py
Comment (WillAyd) [reply scripts/validate_docstrings.py]: All of the calls within the code base are using an instance of the class (save the example in the docstring). With that being the case would prefer to remove the decorator unless there's actually a use case for it; otherwise just adds fluff

**Review by WillAyd — CHANGES_REQUESTED** [review]
One last comment from me then I'll leave to others for review :-)

File: scripts/validate_docstrings.py
Comment (datapythonista) [reply scripts/validate_docstrings.py]: Not a problem at all to make it a regular method, I leave often static methods as regular methods, as it makes almost no difference.

But I don't understand your comment about using an instance of the class. Do you mean that it's not something that we usually use?

The way I see it is that methods of classes in general access the attributes and other methods of the class. As they are encapsulate a set of data and code that somehow works together. Functions in the other hand are somehow independent. In few cases, there are functions that while they do not access anything else from the class, they make more sense inside the class than outside. This is the case here, as `_load_obj` is only used by other methods of the class, and living outside at the same level of the class `Docstring`, IMO makes it look more important in the code than what really is, as conceptually it makes more sense at the same level of the other methods.

Assuming that as a method of the class is a better decision, having it as an staticmethod makes very quick to understand that it's a function inside of a class, and not a method that can access other methods or attributes. I personally find this information very valuable when reading code. Normal methods can access anything in the class, adding an extra complexity in the way that you can't see them as pure input->output units, but that can instead use somehow global variables. A static method doesn't have this complexity, and I know I can expect it to use only whatever parameter it receives, return a result, and not modify anything else. So, it gives a lot of information without having to read its code. The cost is just having the text `@staticmethod` instead of `self`. I think the few characters of difference is a very good price for the amount of information it provides. And I don't think that not using them often in the pandas code itself is a good reason for not using it here. May be we should start using them more there.

Comment (jorisvandenbossche) [reply scripts/validate_docstrings.py]: \+1 on explicitly using staticmethod, if it is actually one

Comment (WillAyd) [reply scripts/validate_docstrings.py]: My point is that I don't see the benefit of just adding things like this without any accompanying test regardless of how trivial the change may seem

Comment (datapythonista) [reply scripts/validate_docstrings.py]: I'm a bit lost on what's the connection between the staticmethod and the testing here.

The benefit of this PR is that makes the code more organized and better. And it implements a way to extract detailed information of the errors in the docstrings. So I can quickly generate the list of functions that fail a specific validation, fix the problem in all them, and start checking in the CI those validations. So the errors don't happen again, and I don't need to waste time on manually reviewing them in all the docstrings PRs.

A huge benefit for me. :)  So would be great to have this merged. Is there any blocker, or can I merge on green?

Comment (WillAyd) [reply scripts/validate_docstrings.py]: I've blocked long enough so if others are fine with it majority rules :-) 

Just generally trying to minimize PRs and keep them directed towards one clear goal at a time, hence the pushback on this one

Comment (datapythonista) [reply scripts/validate_docstrings.py]: hehe, I'm trying to maximize PRs, so hopefully we'll end up in a good equilibrium. ;) After this is merged I'll start creating issues for the points in #20298, so we have better validation of the docstrings.

@jorisvandenbossche happy to merge this?

**Comment by jorisvandenbossche** [comment]
Sorry forgot about this one. Thanks for moving this forward!

**Pull request merged**
